File: dotnet/Lexy/Lexy.Tests/Parser/ExpressionParser/ParseExpressionTestExtensions.cs
using System;
using Lexy.Compiler.Infrastructure;
using Lexy.Compiler.Language.Expressions;
using Lexy.Compiler.Parser;
using Shouldly;

namespace Lexy.Poc.Parser.ExpressionParser;

public static class ParseExpressionTestExtensions
{
    public static Expression ParseExpression(this ScopedServicesTestFixture fixture, string expression)
    {
        var context = fixture.GetService<IParserContext>();
        var tokenizer = fixture.GetService<ITokenizer>();
        var sourceFile = new SourceFile("tests.lexy");
        var line = new Line(0, expression, sourceFile);

        if (!line.Tokenize(tokenizer, context))
            throw new InvalidOperationException("Tokenizing failed: " + context.Logger.ErrorMessages().Format(2));

        var result = ExpressionFactory.Parse(line.Tokens, line);
        result.IsSuccess.ShouldBeTrue(result.ErrorMessage);
        return result.Result;
    }

    public static void ParseExpressionExpectException(this ScopedServicesTestFixture fixture,
        string expression,
        string errorMessage)
    {
        var context = fixture.GetService<IParserContext>();
        var tokenizer = fixture.GetService<ITokenizer>();
        var sourceFile = new SourceFile("tests.lexy");
        var line = new Line(0, expression, sourceFile);

        if (!line.Tokenize(tokenizer, context))
        {
            if (!context.Logger.HasErrorMessage(errorMessage))
                throw new InvalidOperationException(
                    $"Tokenizing failed: {context.Logger.ErrorMessages().Format(2)}");
            return;
        }

        var result = ExpressionFactory.Parse(line.Tokens, line);
        result.IsSuccess.ShouldBeFalse();
        result.ErrorMessage.ShouldBe(errorMessage);
    }
}

File: typescipt/tests/parser/expressionParser/parseExpressionTestExtensions.ts


export class ParseExpressionTestExtensions {
   public static parseExpression(fixture: ScopedServicesTestFixture, expression: string): Expression {
     let context = fixture.GetService<IParserContext>();
     let tokenizer = fixture.GetService<ITokenizer>();
     let sourceFile = new SourceFile(`tests.lexy`);
     let line = new Line(0, expression, sourceFile);

     let tokens = line.Tokenize(tokenizer);
     if (!tokens.state != 'success') {
       throw new Error(`Tokenizing failed: ` + context.logger.errorMessages().Format(2));
     }

     let result = ExpressionFactory.parse(line.tokens, line);
     result.IsSuccess.ShouldBeTrue(result.errorMessage);
     return result.result;
   }

   public static void ParseExpressionExpectException(this ScopedServicesTestFixture fixture,
     string expression,
     string errorMessage) {
     let context = fixture.GetService<IParserContext>();
     let tokenizer = fixture.GetService<ITokenizer>();
     let sourceFile = new SourceFile(`tests.lexy`);
     let line = new Line(0, expression, sourceFile);

     let tokens = line.Tokenize(tokenizer);
     if (!tokens.state != 'success') {
       throw new Error(`Tokenizing failed: ` + context.logger.errorMessages().Format(2));
     }

     let result = ExpressionFactory.parse(line.tokens, line);
     result.IsSuccess.ShouldBeFalse();
     result.errorMessage.ShouldBe(errorMessage);
   }
}
